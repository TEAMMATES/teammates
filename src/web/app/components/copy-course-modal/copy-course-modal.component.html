<div class="modal-header">
  <h5 class="modal-title">
    <div>Copy course</div>
    <small *ngIf="!isCopyFromOtherSession">Copy from: [{{oldCourseId}}] <b>{{oldCourseName}}</b></small>
  </h5>
  <button type="button" class="btn-close" (click)="activeModal.dismiss()"></button>
</div>
<div id="copy-course-modal" class="modal-body">
  <div class="row">
    <div class="col-12">
      <div>
        <label><b>New Course</b></label>
      </div>
      <div class="form-group">
        <label>Course ID:</label>
        <input [class.invalid]="newCourseIdIsConflicting || courseIdEmptyError" id="copy-course-id" type="text" class="form-control" placeholder="e.g. CS3215-2013Semester1"
          [(ngModel)]="newCourseId" [maxlength]="COURSE_ID_MAX_LENGTH" (focus)="this.newCourseIdIsConflicting = false; courseIdEmptyError = false" (ngModelChange)="onCourseIdChange($event)">
        <span>{{ COURSE_ID_MAX_LENGTH - newCourseId.length }} characters left</span>
        <div *ngIf="courseIdEmptyError" class="text-danger">
          <i class="fa fa-exclamation-circle" aria-hidden="true"></i>The field Course ID should not be empty.</div>
      </div>
      <div class="form-group">
        <label>Course Name:</label>
        <input id="copy-course-name" class="form-control" type="text" placeholder="e.g. Software Engineering" [(ngModel)]="newCourseName"
          [maxlength]="COURSE_NAME_MAX_LENGTH" (focus)="courseNameEmptyError = false" (ngModelChange)="onCourseNameChange($event)"/>
        <span>{{ COURSE_NAME_MAX_LENGTH - newCourseName.length }} characters left</span>
        <div *ngIf="courseNameEmptyError" class="text-danger">
          <i class="fa fa-exclamation-circle" aria-hidden="true"></i>The field Course Name should not be empty.</div>
      </div>
      <div class="form-group">
        <label class="ngb-tooltip-class">
          <span ngbTooltip="Note that you can only create a new course under an institute in which you are already a course co-owner.">
            Course institute:
          </span>
        </label>
        <div class="input-group">
          <select id="copy-course-institute" class="form-control form-select" [(ngModel)]="newCourseInstitute">
            <option *ngFor="let institute of institutes" [value]="institute">{{ institute }}</option>
          </select>
        </div>
      </div>
      <div class="form-group">
        <label class="ngb-tooltip-class"><span ngbTooltip="You should not need to change this as it is auto-detected based on your device settings. TEAMMATES automatically adjusts
          to match the current time offset in your area, including clock changes due to daylight saving time.">Time Zone:</span></label>
        <div class="input-group">
          <select id="copy-time-zone" class="form-control form-select" [(ngModel)]="newTimezone">
            <option *ngFor="let timezone of timezones" [value]="timezone.id">{{ timezone.id }} ({{ timezone.offset }})</option>
          </select>
          <span class="input-group-btn" style="margin-left: 5px;">
            <button class="btn btn-primary btn-margin-left" (click)="onAutoDetectTimezone()">Auto Detect</button>
          </span>
        </div>
      </div>
      <div class="form-group" *ngIf="isCopyFromOtherSession">
        <label>Course to copy from:</label>
        <div class="input-group">
          <select id="course-to-copy-from" class="form-control form-select" (change)="onSelectCourseChange()" [(ngModel)]="oldCourseId">
            <option *ngFor="let course of activeCourses" [value]="course.courseId">{{ course.courseId }}</option>
          </select>
        </div>
      </div>
      <div *ngIf="oldCourseId">
        <div *ngIf="courseToFeedbackSession[oldCourseId]?.length > 0; else noSessionsTemplate">
          <div class="form-check">
            <input type="checkbox" name="copyCourseChooseSessions" class="form-check-input"
              [checked]="selectedFeedbackSessions.size === courseToFeedbackSession[oldCourseId].length" (click)="toggleSelectionForAll()">
            <span><b>Sessions to copy:</b></span>
          </div>
          <hr class="solid-divider">
          <div class="form-check" *ngFor="let session of courseToFeedbackSession[oldCourseId]">
            <label class="form-check-label">
              <input type="checkbox" name="copyCourseChooseSessions" class="form-check-input"
                [checked]="selectedFeedbackSessions.has(session)" (click)="toggleSelection(session)">
              <span>{{ session.feedbackSessionName }}</span>
            </label>
          </div>
        </div>
        <ng-template #noSessionsTemplate>
          <i>This course has no feedback sessions</i>
        </ng-template>
      </div>
    </div>
  </div>
</div>
<div class="modal-footer">
  <button type="button" class="btn btn-light" (click)="activeModal.dismiss()">Cancel</button>
  <button id="btn-confirm-copy-course" type="button" class="btn btn-primary" (click)="copy()"
    [disabled]="!newCourseId || !newCourseName || !newCourseInstitute">Copy</button>
</div>
