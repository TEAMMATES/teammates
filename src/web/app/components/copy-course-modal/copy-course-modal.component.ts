import { Component, EventEmitter, Input, OnInit } from '@angular/core';
import { NgbActiveModal } from '@ng-bootstrap/ng-bootstrap';
import { CopyCourseModalResult } from './copy-course-modal-model';
import { StatusMessageService } from '../../../services/status-message.service';
import { TimezoneService } from '../../../services/timezone.service';
import { Course, FeedbackSession } from '../../../types/api-output';
import { COURSE_ID_MAX_LENGTH, COURSE_NAME_MAX_LENGTH } from '../../../types/field-validator';

interface Timezone {
  id: string;
  offset: string;
}

const zeroPad: (num: number) => string = (num: number) => String(num).padStart(2, '0');

/**
 * Copy course modal.
 */
@Component({
  selector: 'tm-copy-course-modal',
  templateUrl: './copy-course-modal.component.html',
  styleUrls: ['./copy-course-modal.component.scss'],
})
export class CopyCourseModalComponent implements OnInit {

  // const
  readonly COURSE_ID_MAX_LENGTH: number = COURSE_ID_MAX_LENGTH;
  readonly COURSE_NAME_MAX_LENGTH: number = COURSE_NAME_MAX_LENGTH;

  @Input()
  courseToFeedbackSession: Record<string, FeedbackSession[]> = {};

  @Input()
  activeCourses: Course[] = [];

  @Input()
  allCourses: Course[] = [];

  fetchFeedbackSessionsEvent: EventEmitter<string> = new EventEmitter<string>();

  isCopyFromOtherSession: boolean = false;
  newCourseIdIsConflicting: boolean = false;
  courseIdEmptyError: boolean = false;
  courseNameEmptyError: boolean = false;
  institutes: string[] = [];
  timezones: Timezone[] = [];
  newTimezone: string = '';
  newCourseId: string = '';
  newCourseName: string = '';
  newCourseInstitute: string = '';
  oldCourseId: string = '';
  oldCourseName: string = '';

  selectedFeedbackSessions: Set<FeedbackSession> = new Set<FeedbackSession>();

  constructor(public activeModal: NgbActiveModal,
              private statusMessageService: StatusMessageService,
              private timezoneService: TimezoneService) {}

  ngOnInit(): void {
    this.timezones = Object.entries(this.timezoneService.getTzOffsets())
      .map(([id, offset]: [string, number]) => {
        const hourOffset: number = Math.floor(Math.abs(offset) / 60);
        const minOffset: number = Math.abs(offset) % 60;
        const sign: string = offset < 0 ? '-' : '+';
        return { id, offset: offset === 0 ? 'UTC' : `UTC ${sign}${zeroPad(hourOffset)}:${zeroPad(minOffset)}` };
      });
    this.institutes = Array.from(new Set(this.allCourses.map((course: Course) => course.institute)));
    if (this.institutes.length) {
      this.newCourseInstitute = this.institutes[0];
    }
    this.newTimezone = this.timezoneService.guessTimezone();

    this.courseIdEmptyError = false;
    this.courseNameEmptyError = false;
  }

  /**
   * Fires the copy event.
   */
  copy(): void {
    this.courseIdEmptyError = !this.newCourseId;
    this.courseNameEmptyError = !this.newCourseName;

    if (this.courseIdEmptyError || this.courseNameEmptyError) {
      return;
    }

    this.newCourseIdIsConflicting = this.allCourses
      .filter((course: Course) => course.courseId === this.newCourseId).length > 0;
    if (this.newCourseIdIsConflicting) {
      this.statusMessageService.showErrorToast(
        `The course ID ${this.newCourseId} already exists.`);
      return;
    }

    const result: CopyCourseModalResult = {
      newCourseId: this.newCourseId,
      newCourseName: this.newCourseName,
      newCourseInstitute: this.newCourseInstitute,
      oldCourseId: this.oldCourseId,
      newTimeZone: this.newTimezone,
      selectedFeedbackSessionList: this.selectedFeedbackSessions,
      totalNumberOfSessions: this.selectedFeedbackSessions.size,
    };

    this.activeModal.close(result);
  }
  /**
    * Real time check user input, triggered by ngModelChange
    */
  onCourseIdChange(value: string): void {
    this.courseIdEmptyError = !value.trim();
   }

  onCourseNameChange(value: string): void {
    this.courseNameEmptyError = !value.trim();
  }

  /**
   * Toggles selection of a feedback session.
   */
  toggleSelection(session: FeedbackSession): void {
    if (this.selectedFeedbackSessions.has(session)) {
      this.selectedFeedbackSessions.delete(session);
    } else {
      this.selectedFeedbackSessions.add(session);
    }
  }

  /**
   * Select all sessions or clear all sessions
   */
  toggleSelectionForAll(): void {
    if (this.selectedFeedbackSessions.size === this.courseToFeedbackSession[this.oldCourseId]?.length) {
      this.selectedFeedbackSessions.clear();
    } else {
      this.selectedFeedbackSessions = new Set(this.courseToFeedbackSession[this.oldCourseId]);
    }
  }

  /**
   * Auto-detects timezone for instructor.
   */
  onAutoDetectTimezone(): void {
    this.newTimezone = this.timezoneService.guessTimezone();
  }

  /**
   * Clears all selected feedback sessions of previously chosen course.
   */
  onSelectCourseChange(): void {
    this.selectedFeedbackSessions.clear();

    if (!this.courseToFeedbackSession[this.oldCourseId]) {
      this.fetchFeedbackSessionsEvent.emit(this.oldCourseId);
    }
  }
}
