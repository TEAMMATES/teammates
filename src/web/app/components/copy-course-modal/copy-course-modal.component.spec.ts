import { HttpClientTestingModule } from '@angular/common/http/testing';
import { ComponentFixture, TestBed, waitForAsync } from '@angular/core/testing';
import { FormsModule } from '@angular/forms';
import { By } from '@angular/platform-browser';
import { NgbActiveModal } from '@ng-bootstrap/ng-bootstrap';
import SpyInstance = jest.SpyInstance;
import { CopyCourseModalComponent } from './copy-course-modal.component';
import { StatusMessageService } from '../../../services/status-message.service';
import { TimezoneService } from '../../../services/timezone.service';
import {
  Course,
  FeedbackSession,
  FeedbackSessionPublishStatus,
  FeedbackSessionSubmissionStatus,
  ResponseVisibleSetting,
  SessionVisibleSetting,
} from '../../../types/api-output';

describe('CopyCourseModalComponent', () => {
  let component: CopyCourseModalComponent;
  let ngbActiveModal: NgbActiveModal;
  let timezoneService: TimezoneService;
  let statusMessageService: StatusMessageService;
  let fixture: ComponentFixture<CopyCourseModalComponent>;

  beforeEach(waitForAsync(() => {
    TestBed.configureTestingModule({
      declarations: [CopyCourseModalComponent],
      imports: [
        HttpClientTestingModule,
        FormsModule,
      ],
      providers: [NgbActiveModal, TimezoneService, StatusMessageService],
    })
    .compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(CopyCourseModalComponent);
    timezoneService = TestBed.inject(TimezoneService);
    statusMessageService = TestBed.inject(StatusMessageService);
    ngbActiveModal = TestBed.inject(NgbActiveModal);
    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should snap with default fields', () => {
    component.timezones = [{
      id: 'Asia/Singapore',
      offset: 'UTC +08:00',
    }, {
      id: 'UTC',
      offset: 'UTC',
    }];
    fixture.detectChanges();
    expect(fixture).toMatchSnapshot();
  });

  it('should snap with some course id', () => {
    component.timezones = [{
      id: 'Asia/Singapore',
      offset: 'UTC +08:00',
    }, {
      id: 'UTC',
      offset: 'UTC',
    }];
    component.newCourseId = 'Test02';
    fixture.detectChanges();
    expect(fixture).toMatchSnapshot();
  });

  it('should snap when copying from other sessions', () => {
    component.timezones = [{
      id: 'Asia/Singapore',
      offset: 'UTC +08:00',
    }, {
      id: 'UTC',
      offset: 'UTC',
    }];
    component.isCopyFromOtherSession = true;
    fixture.detectChanges();
    expect(fixture).toMatchSnapshot();
  });

  it('should enable copy button after new courseId is provided', () => {
    component.newCourseId = 'Test02';
    component.newCourseName = 'TestName02';
    component.newCourseInstitute = 'Test institute';
    fixture.detectChanges();
    const copyButton: any = fixture.debugElement.query(By.css('#btn-confirm-copy-course'));
    expect(copyButton.nativeElement.disabled).toBeFalsy();
  });

  it('should disable copy if courseId is empty', () => {
    component.newCourseId = '';
    component.newCourseName = 'TestName02';
    fixture.detectChanges();
    const copyButton: any = fixture.debugElement.query(By.css('#btn-confirm-copy-course'));
    expect(copyButton.nativeElement.disabled).toBeTruthy();
  });

  it('should toggle selection', () => {
    const testFeedbackSession: FeedbackSession = {
      courseId: 'testId',
      timeZone: 'Asia/Singapore',
      feedbackSessionName: 'Test Session',
      instructions: 'Instructions',
      submissionStartTimestamp: 1000000000000,
      submissionEndTimestamp: 1500000000000,
      gracePeriod: 0,
      sessionVisibleSetting: SessionVisibleSetting.AT_OPEN,
      responseVisibleSetting: ResponseVisibleSetting.AT_VISIBLE,
      submissionStatus: FeedbackSessionSubmissionStatus.OPEN,
      publishStatus: FeedbackSessionPublishStatus.PUBLISHED,
      isClosingSoonEmailEnabled: true,
      isPublishedEmailEnabled: true,
      createdAtTimestamp: 0,
      studentDeadlines: {},
      instructorDeadlines: {},
    };
    component.selectedFeedbackSessions.add(testFeedbackSession);
    fixture.detectChanges();
    component.toggleSelection(testFeedbackSession);
    expect(component.selectedFeedbackSessions.has(testFeedbackSession)).toEqual(false);
  });

  it('should set newCourseInstitute if institutes array is not empty', () => {
    const testCourses: Course[] = [
      {
        courseId: 'testId1',
        courseName: 'testCourse1',
        timeZone: 'Asia/Singapore',
        institute: 'Institute 1',
        creationTimestamp: 1000000000000,
        deletionTimestamp: 1500000000000,
      },
      {
        courseId: 'testId2',
        courseName: 'testCourse2',
        timeZone: 'Asia/Singapore',
        institute: 'Institute 2',
        creationTimestamp: 1000000000000,
        deletionTimestamp: 1500000000000,
      },
    ];
    component.allCourses = testCourses;
    component.ngOnInit();

    expect(component.newCourseInstitute).toBe('Institute 1');
  });

  it('should set the timezones property correctly', () => {
    const mockTzOffsets = {
      Tz1: 120, // UTC+02:00
      Tz2: -180, // UTC-03:00
      Tz3: 0,
    };

    jest.spyOn(timezoneService, 'getTzOffsets').mockReturnValue(mockTzOffsets);

    component.ngOnInit();

    expect(component.timezones).toEqual([
      { id: 'Tz1', offset: 'UTC +02:00' },
      { id: 'Tz2', offset: 'UTC -03:00' },
      { id: 'Tz3', offset: 'UTC' },
    ]);
  });

  it('should call showErrorToast when copying with no new courseId and name', () => {
    const spyStatusMessageService: SpyInstance = jest.spyOn(statusMessageService, 'showErrorToast');
    component.copy();
    expect(spyStatusMessageService)
      .toHaveBeenCalledWith('Please make sure you have filled in both Course ID and Name before adding the course!');
  });

  it('should call showErrorToast when newCourseId is a duplicate', () => {
    const testCourses: Course[] = [
      {
        courseId: 'testId1',
        courseName: 'testCourse1',
        timeZone: 'Asia/Singapore',
        institute: 'Institute 1',
        creationTimestamp: 1000000000000,
        deletionTimestamp: 1500000000000,
      },
      {
        courseId: 'testId2',
        courseName: 'testCourse2',
        timeZone: 'Asia/Singapore',
        institute: 'Institute 2',
        creationTimestamp: 1000000000000,
        deletionTimestamp: 1500000000000,
      },
    ];
    const spyStatusMessageService: SpyInstance = jest.spyOn(statusMessageService, 'showErrorToast');
    component.allCourses = testCourses;
    component.newCourseId = 'testId1';
    component.newCourseName = 'testName';
    component.copy();
    expect(spyStatusMessageService).toHaveBeenCalledWith('The course ID testId1 already exists.');
  });

  it('should call close with the result when succesfully copying a course', () => {
    const activeModalSpy: SpyInstance = jest.spyOn(ngbActiveModal, 'close');
    component.newCourseId = 'testId2';
    component.newCourseName = 'testName';
    component.newTimezone = 'Asia/Singapore';
    component.copy();
    expect(activeModalSpy).toHaveBeenCalledWith({
      newCourseId: 'testId2',
      newCourseName: 'testName',
      newCourseInstitute: '',
      newTimeZone: 'Asia/Singapore',
      oldCourseId: '',
      selectedFeedbackSessionList: new Set(),
      totalNumberOfSessions: 0,
    });
  });

  it('should toggle selection of all feedbacksessions correctly', () => {
    const testFeedbackSession1: FeedbackSession = {
      courseId: 'testId',
      timeZone: 'Asia/Singapore',
      feedbackSessionName: 'Test Session 1',
      instructions: 'Instructions',
      submissionStartTimestamp: 1000000000000,
      submissionEndTimestamp: 1500000000000,
      gracePeriod: 0,
      sessionVisibleSetting: SessionVisibleSetting.AT_OPEN,
      responseVisibleSetting: ResponseVisibleSetting.AT_VISIBLE,
      submissionStatus: FeedbackSessionSubmissionStatus.OPEN,
      publishStatus: FeedbackSessionPublishStatus.PUBLISHED,
      isClosingSoonEmailEnabled: true,
      isPublishedEmailEnabled: true,
      createdAtTimestamp: 0,
      studentDeadlines: {},
      instructorDeadlines: {},
    };
    const testFeedbackSession2: FeedbackSession = {
      courseId: 'testId',
      timeZone: 'Asia/Singapore',
      feedbackSessionName: 'Test Session 2',
      instructions: 'Instructions',
      submissionStartTimestamp: 1000000000000,
      submissionEndTimestamp: 1500000000000,
      gracePeriod: 0,
      sessionVisibleSetting: SessionVisibleSetting.AT_OPEN,
      responseVisibleSetting: ResponseVisibleSetting.AT_VISIBLE,
      submissionStatus: FeedbackSessionSubmissionStatus.OPEN,
      publishStatus: FeedbackSessionPublishStatus.PUBLISHED,
      isClosingSoonEmailEnabled: true,
      isPublishedEmailEnabled: true,
      createdAtTimestamp: 0,
      studentDeadlines: {},
      instructorDeadlines: {},
    };
    const testFeedbackSessions = [testFeedbackSession1, testFeedbackSession2];
    component.courseToFeedbackSession = {
      testCourseId: testFeedbackSessions,
    };
    component.oldCourseId = 'testCourseId';

    // selectedFeedbackSessions is empty
    component.toggleSelectionForAll();
    expect(component.selectedFeedbackSessions).toEqual(new Set(testFeedbackSessions));
    // selectedFeedbackSessions has all the feedback sessions of the old course
    component.toggleSelectionForAll();
    expect(component.selectedFeedbackSessions.size).toBe(0);
    // selectedFeedbackSessions is partially filled
    component.selectedFeedbackSessions = new Set([testFeedbackSession1]);
    component.toggleSelectionForAll();
    expect(component.selectedFeedbackSessions).toEqual(new Set(testFeedbackSessions));
  });

  it('should set newTimezone correctly when onAutoDetectTimezone is called', () => {
    const mockTimezone = 'timezone';
    jest.spyOn(timezoneService, 'guessTimezone').mockReturnValue(mockTimezone);
    component.onAutoDetectTimezone();
    expect(component.newTimezone).toBe(mockTimezone);
  });

  it('should clear selectedFeedbackSessions when onSelectCourseChange is called', () => {
    const testFeedbackSession: FeedbackSession = {
      courseId: 'testId2',
      timeZone: 'Asia/Singapore',
      feedbackSessionName: 'Test Session 2',
      instructions: 'Instructions',
      submissionStartTimestamp: 1000000000000,
      submissionEndTimestamp: 1500000000000,
      gracePeriod: 0,
      sessionVisibleSetting: SessionVisibleSetting.AT_OPEN,
      responseVisibleSetting: ResponseVisibleSetting.AT_VISIBLE,
      submissionStatus: FeedbackSessionSubmissionStatus.OPEN,
      publishStatus: FeedbackSessionPublishStatus.PUBLISHED,
      isClosingSoonEmailEnabled: true,
      isPublishedEmailEnabled: true,
      createdAtTimestamp: 0,
      studentDeadlines: {},
      instructorDeadlines: {},
    };
    component.oldCourseId = 'testId1';
    component.courseToFeedbackSession = {
      testId2: [testFeedbackSession],
    };
    component.selectedFeedbackSessions = new Set([testFeedbackSession]);
    let emittedCourseId;
    component.fetchFeedbackSessionsEvent.subscribe((emittedValue) => { emittedCourseId = emittedValue; });
    component.onSelectCourseChange();
    expect(component.selectedFeedbackSessions.size).toBe(0);
    expect(emittedCourseId).toEqual('testId1');
  });
});
