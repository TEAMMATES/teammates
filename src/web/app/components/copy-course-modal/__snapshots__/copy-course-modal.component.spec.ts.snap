// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CopyCourseModalComponent should snap when copying from other sessions 1`] = `
<tm-copy-course-modal
  COURSE_ID_MAX_LENGTH={[Function Number]}
  COURSE_NAME_MAX_LENGTH={[Function Number]}
  activeCourses={[Function Array]}
  activeModal={[Function NgbActiveModal]}
  allCourses={[Function Array]}
  courseIdEmptyError="false"
  courseNameEmptyError="false"
  courseToFeedbackSession={[Function Object]}
  fetchFeedbackSessionsEvent={[Function EventEmitter_]}
  institutes={[Function Array]}
  isCopyFromOtherSession={[Function Boolean]}
  newCourseId=""
  newCourseIdIsConflicting="false"
  newCourseInstitute=""
  newCourseName=""
  newTimezone={[Function String]}
  oldCourseId=""
  oldCourseName=""
  selectedFeedbackSessions={[Function Set]}
  statusMessageService={[Function StatusMessageService]}
  timezoneService={[Function TimezoneService]}
  timezones={[Function Array]}
>
  <div
    class="modal-header"
  >
    <h5
      class="modal-title"
    >
      <div>
        Copy course
      </div>
    </h5>
    <button
      class="btn-close"
      type="button"
    />
  </div><div
    class="modal-body"
    id="copy-course-modal"
  >
    <div
      class="row"
    >
      <div
        class="col-12"
      >
        <div>
          <label>
            <b>
              New Course
            </b>
          </label>
        </div>
        <div
          class="form-group"
        >
          <label>
            Course ID:
          </label>
          <input
            class="form-control ng-untouched ng-pristine ng-valid"
            id="copy-course-id"
            maxlength="64"
            placeholder="e.g. CS3215-2013Semester1"
            type="text"
          />
          <span>
            64 characters left
          </span>
        </div>
        <div
          class="form-group"
        >
          <label>
            Course Name:
          </label>
          <input
            class="form-control ng-untouched ng-pristine ng-valid"
            id="copy-course-name"
            maxlength="80"
            placeholder="e.g. Software Engineering"
            type="text"
          />
          <span>
            80 characters left
          </span>
        </div>
        <div
          class="form-group"
        >
          <label
            class="ngb-tooltip-class"
          >
            <span
              ngbtooltip="Note that you can only create a new course under an institute in which you are already a course co-owner."
            >
               Course institute: 
            </span>
          </label>
          <div
            class="input-group"
          >
            <select
              class="form-control form-select ng-untouched ng-pristine ng-valid"
              id="copy-course-institute"
            >
            </select>
          </div>
        </div>
        <div
          class="form-group"
        >
          <label
            class="ngb-tooltip-class"
          >
            <span
              ngbtooltip="You should not need to change this as it is auto-detected based on your device settings. TEAMMATES automatically adjusts
            to match the current time offset in your area, including clock changes due to daylight saving time."
            >
              Time Zone:
            </span>
          </label>
          <div
            class="input-group"
          >
            <select
              class="form-control form-select ng-untouched ng-pristine ng-valid"
              id="copy-time-zone"
            >
              <option
                value="Asia/Singapore"
              >
                Asia/Singapore (UTC +08:00)
              </option>
              <option
                value="UTC"
              >
                UTC (UTC)
              </option>
            </select>
            <span
              class="input-group-btn"
              style="margin-left: 5px;"
            >
              <button
                class="btn btn-primary btn-margin-left"
              >
                Auto Detect
              </button>
            </span>
          </div>
        </div>
        <div
          class="form-group"
        >
          <label>
            Course to copy from:
          </label>
          <div
            class="input-group"
          >
            <select
              class="form-control form-select ng-untouched ng-pristine ng-valid"
              id="course-to-copy-from"
            >
            </select>
          </div>
        </div>
      </div>
    </div>
  </div><div
    class="modal-footer"
  >
    <button
      class="btn btn-light"
      type="button"
    >
      Cancel
    </button>
    <button
      class="btn btn-primary"
      disabled=""
      id="btn-confirm-copy-course"
      type="button"
    >
      Copy
    </button>
  </div>
</tm-copy-course-modal>
`;

exports[`CopyCourseModalComponent should snap with default fields 1`] = `
<tm-copy-course-modal
  COURSE_ID_MAX_LENGTH={[Function Number]}
  COURSE_NAME_MAX_LENGTH={[Function Number]}
  activeCourses={[Function Array]}
  activeModal={[Function NgbActiveModal]}
  allCourses={[Function Array]}
  courseIdEmptyError="false"
  courseNameEmptyError="false"
  courseToFeedbackSession={[Function Object]}
  fetchFeedbackSessionsEvent={[Function EventEmitter_]}
  institutes={[Function Array]}
  isCopyFromOtherSession="false"
  newCourseId=""
  newCourseIdIsConflicting="false"
  newCourseInstitute=""
  newCourseName=""
  newTimezone={[Function String]}
  oldCourseId=""
  oldCourseName=""
  selectedFeedbackSessions={[Function Set]}
  statusMessageService={[Function StatusMessageService]}
  timezoneService={[Function TimezoneService]}
  timezones={[Function Array]}
>
  <div
    class="modal-header"
  >
    <h5
      class="modal-title"
    >
      <div>
        Copy course
      </div>
      <small>
        Copy from: [] 
        <b>
        </b>
      </small>
    </h5>
    <button
      class="btn-close"
      type="button"
    />
  </div><div
    class="modal-body"
    id="copy-course-modal"
  >
    <div
      class="row"
    >
      <div
        class="col-12"
      >
        <div>
          <label>
            <b>
              New Course
            </b>
          </label>
        </div>
        <div
          class="form-group"
        >
          <label>
            Course ID:
          </label>
          <input
            class="form-control ng-untouched ng-pristine ng-valid"
            id="copy-course-id"
            maxlength="64"
            placeholder="e.g. CS3215-2013Semester1"
            type="text"
          />
          <span>
            64 characters left
          </span>
        </div>
        <div
          class="form-group"
        >
          <label>
            Course Name:
          </label>
          <input
            class="form-control ng-untouched ng-pristine ng-valid"
            id="copy-course-name"
            maxlength="80"
            placeholder="e.g. Software Engineering"
            type="text"
          />
          <span>
            80 characters left
          </span>
        </div>
        <div
          class="form-group"
        >
          <label
            class="ngb-tooltip-class"
          >
            <span
              ngbtooltip="Note that you can only create a new course under an institute in which you are already a course co-owner."
            >
               Course institute: 
            </span>
          </label>
          <div
            class="input-group"
          >
            <select
              class="form-control form-select ng-untouched ng-pristine ng-valid"
              id="copy-course-institute"
            >
            </select>
          </div>
        </div>
        <div
          class="form-group"
        >
          <label
            class="ngb-tooltip-class"
          >
            <span
              ngbtooltip="You should not need to change this as it is auto-detected based on your device settings. TEAMMATES automatically adjusts
            to match the current time offset in your area, including clock changes due to daylight saving time."
            >
              Time Zone:
            </span>
          </label>
          <div
            class="input-group"
          >
            <select
              class="form-control form-select ng-untouched ng-pristine ng-valid"
              id="copy-time-zone"
            >
              <option
                value="Asia/Singapore"
              >
                Asia/Singapore (UTC +08:00)
              </option>
              <option
                value="UTC"
              >
                UTC (UTC)
              </option>
            </select>
            <span
              class="input-group-btn"
              style="margin-left: 5px;"
            >
              <button
                class="btn btn-primary btn-margin-left"
              >
                Auto Detect
              </button>
            </span>
          </div>
        </div>
      </div>
    </div>
  </div><div
    class="modal-footer"
  >
    <button
      class="btn btn-light"
      type="button"
    >
      Cancel
    </button>
    <button
      class="btn btn-primary"
      disabled=""
      id="btn-confirm-copy-course"
      type="button"
    >
      Copy
    </button>
  </div>
</tm-copy-course-modal>
`;

exports[`CopyCourseModalComponent should snap with some course id 1`] = `
<tm-copy-course-modal
  COURSE_ID_MAX_LENGTH={[Function Number]}
  COURSE_NAME_MAX_LENGTH={[Function Number]}
  activeCourses={[Function Array]}
  activeModal={[Function NgbActiveModal]}
  allCourses={[Function Array]}
  courseIdEmptyError="false"
  courseNameEmptyError="false"
  courseToFeedbackSession={[Function Object]}
  fetchFeedbackSessionsEvent={[Function EventEmitter_]}
  institutes={[Function Array]}
  isCopyFromOtherSession="false"
  newCourseId={[Function String]}
  newCourseIdIsConflicting="false"
  newCourseInstitute=""
  newCourseName=""
  newTimezone={[Function String]}
  oldCourseId=""
  oldCourseName=""
  selectedFeedbackSessions={[Function Set]}
  statusMessageService={[Function StatusMessageService]}
  timezoneService={[Function TimezoneService]}
  timezones={[Function Array]}
>
  <div
    class="modal-header"
  >
    <h5
      class="modal-title"
    >
      <div>
        Copy course
      </div>
      <small>
        Copy from: [] 
        <b>
        </b>
      </small>
    </h5>
    <button
      class="btn-close"
      type="button"
    />
  </div><div
    class="modal-body"
    id="copy-course-modal"
  >
    <div
      class="row"
    >
      <div
        class="col-12"
      >
        <div>
          <label>
            <b>
              New Course
            </b>
          </label>
        </div>
        <div
          class="form-group"
        >
          <label>
            Course ID:
          </label>
          <input
            class="form-control ng-untouched ng-pristine ng-valid"
            id="copy-course-id"
            maxlength="64"
            placeholder="e.g. CS3215-2013Semester1"
            type="text"
          />
          <span>
            58 characters left
          </span>
        </div>
        <div
          class="form-group"
        >
          <label>
            Course Name:
          </label>
          <input
            class="form-control ng-untouched ng-pristine ng-valid"
            id="copy-course-name"
            maxlength="80"
            placeholder="e.g. Software Engineering"
            type="text"
          />
          <span>
            80 characters left
          </span>
        </div>
        <div
          class="form-group"
        >
          <label
            class="ngb-tooltip-class"
          >
            <span
              ngbtooltip="Note that you can only create a new course under an institute in which you are already a course co-owner."
            >
               Course institute: 
            </span>
          </label>
          <div
            class="input-group"
          >
            <select
              class="form-control form-select ng-untouched ng-pristine ng-valid"
              id="copy-course-institute"
            >
            </select>
          </div>
        </div>
        <div
          class="form-group"
        >
          <label
            class="ngb-tooltip-class"
          >
            <span
              ngbtooltip="You should not need to change this as it is auto-detected based on your device settings. TEAMMATES automatically adjusts
            to match the current time offset in your area, including clock changes due to daylight saving time."
            >
              Time Zone:
            </span>
          </label>
          <div
            class="input-group"
          >
            <select
              class="form-control form-select ng-untouched ng-pristine ng-valid"
              id="copy-time-zone"
            >
              <option
                value="Asia/Singapore"
              >
                Asia/Singapore (UTC +08:00)
              </option>
              <option
                value="UTC"
              >
                UTC (UTC)
              </option>
            </select>
            <span
              class="input-group-btn"
              style="margin-left: 5px;"
            >
              <button
                class="btn btn-primary btn-margin-left"
              >
                Auto Detect
              </button>
            </span>
          </div>
        </div>
      </div>
    </div>
  </div><div
    class="modal-footer"
  >
    <button
      class="btn btn-light"
      type="button"
    >
      Cancel
    </button>
    <button
      class="btn btn-primary"
      disabled=""
      id="btn-confirm-copy-course"
      type="button"
    >
      Copy
    </button>
  </div>
</tm-copy-course-modal>
`;
